aggregator/src/aggregator.rs

```diff
@@ -221,24 +245,79 @@ impl<C: Clock> Aggregator<C> {
             upload_decrypt_failure_counter,
             upload_decode_failure_counter,
             aggregate_step_failure_counter,
+            global_hpke_configs: Mutex::new(GlobalHpkeConfigCache::new()),
         }
     }
 
     async fn handle_hpke_config(
         &self,
         task_id_base64: Option<&[u8]>,
     ) -> Result<HpkeConfigList, Error> {
-        // Task ID is optional in an HPKE config request, but Janus requires it.
-        // https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-4.3.1
-        let task_id_base64 = task_id_base64.ok_or(Error::MissingTaskId)?;
-
-        let task_id_bytes = URL_SAFE_NO_PAD
-            .decode(task_id_base64)
-            .map_err(|_| Error::UnrecognizedMessage(None, "task_id"))?;
-        let task_id = TaskId::get_decoded(&task_id_bytes)
-            .map_err(|_| Error::UnrecognizedMessage(None, "task_id"))?;
-        let task_aggregator = self.task_aggregator_for(&task_id).await?;
-        Ok(task_aggregator.handle_hpke_config())
+        match task_id_base64 {
+            Some(task_id_base64) => {
+                let task_id_bytes = URL_SAFE_NO_PAD
+                    .decode(task_id_base64)
+                    .map_err(|_| Error::UnrecognizedMessage(None, "task_id"))?;
+                let task_id = TaskId::get_decoded(&task_id_bytes)
+                    .map_err(|_| Error::UnrecognizedMessage(None, "task_id"))?;
+                let task_aggregator = self.task_aggregator_for(&task_id).await?;
+                Ok(task_aggregator.handle_hpke_config())
+            }
+            None => {
+                let configs = self.get_global_hpke_configs().await?;
+                if configs.is_empty() {
+                    if self.cfg.taskprov_config.enabled {
+                        // A global HPKE configuration is only _required_ when taskprov
+                        // is enabled.
+                        Err(Error::Internal(
+                            "this server is missing its global HPKE config".into(),
+                        ))
+                    } else {
+                        Err(Error::MissingTaskId)
+                    }
+                } else {
+                    Ok(HpkeConfigList::new(configs))
+                }
+            }
+        }
+    }
+
+    /// Get active and non-expired keys from the database.
+    async fn get_global_hpke_configs(&self) -> Result<Vec<HpkeConfig>, Error> {
+        let mut global_hpke_configs = self.global_hpke_configs.lock().await;
+
+        // Slow path: the cache has expired and needs to be refreshed.
+        let now = self.clock.now();
+        if now
+            > global_hpke_configs
+                .last_updated
+                .add(&GlobalHpkeConfigCache::REFRESH_INTERVAL)?
+        {
+            let configs = self
+                .datastore
+                .run_tx_with_name("refresh_global_hpke_configs", |tx| {
```

Comment:
We should not run a DB operation while holding the lock; this will induce unnecessary/artificial latency between different readers of the HPKE configs.

Instead, I suggest:
* On creation of the `GlobalHpkeConfigCache`, start a "refresh" task which periodically refreshes the global HPKE configs (with the first refresh happening immediately).
* On dropping the `GlobalHpkeConfigCache`, stop the task.
* To retrieve the global HPKE configs, use one of the following strategies (see https://docs.rs/tokio/latest/tokio/sync/index.html):
  * an `mpsc::Sender<oneshot::Sender<Arc<Vec<HpkeConfig>>>>` to allow the reader to send a message to the refresh task, which it will respond to by sending a (clone of the `Arc` of) the current `HpkeConfig`s.
  * use a `Watch<Option<Arc<Vec<HpkeConfig>>>>`[1] to store the value -- this works great except that you must have an initial value, so we'd need to somehow block readers from trying to read before the initial HPKE config refresh is finished. One way to do this would be to use an `RwLock` which is initially write-locked; have the refresher task release the write-lock once a value is available in the `Watch`, then have the readers acquire & release a read lock before attempting to read the `Watch` value.

With this strategy, no reader will ever wait after the first time the refresh is completed.

[1] lol @ this type name

Reply:
Would it be sufficient to instead drop the lock before querying the database and reacquire the lock before updating the cache? That way only one `/hpke_config` call will be blocked to update the cache.

This pattern we use already in our task cache https://github.com/divviup/janus/blob/main/aggregator/src/aggregator.rs#L428

Reply:
Well wait, that won't quite work. An updating reader could be in the middle of a tx while another reader queries the cache again. That other reader will move into its tx too, thus duplicating database work.

Isn't this a problem too for the task cache, linked above?

Reply:
The task cache has the same problem, but it's "lesser" since the task is only retrieved once (per process). The global HPKE config retrieval happens once an hour (per process), so the extra DB hits will recur once an hour.

Still, IMO your suggestion is likely an acceptable middle ground/first pass -- I really just want to avoid doing IO while holding a lock.

If you go this route, would you mind filing an issue to fix up both of these? I think the fix would be pretty small but concurrency code is always tricky.

Reply:
> The task cache has the same problem, but it's "lesser" since the task is only retrieved once (per process). The global HPKE config retrieval happens once an hour (per process), so the extra DB hits will recur once an hour.

:+1:

> acceptable middle ground/first pass

I've put together a small isolated experiment to try various caching methods, and I think the two lock method I just suggested is the least effective. If we assume 1000 QPS and 100ms database roundtrip, then that's an additional 100 hits to the database. Those tasks also have wait for the full 100ms roundtrip each, compared to the single lock where blocked requests have less time to wait as the one tx approaches completion.

Here's that harness https://gist.github.com/inahga/3531c7341346b3224d99059cf80928b0, if interested or want to experiment.

I'll use this harness as well to try the two sync approaches you suggested. I _suspect_ that the `Watch` method won't be any better because it internally uses a RwLock on its value, which IIUC will block readers when a write lock is queued. Either way this is a good exercise since I've not delved into tokio's sync primitives very much yet.

---

The radical suggestion would also be to drop the cache and just load all advertiseable keys at runtime. I don't like this though because it means key rotation involves two application restarts since a key state change requires restart.

Reply:
I'll also note that regardless of the caching method used, I believe now a background task for refreshing is strictly necessary. With the refresh on every `get`, the first request after the refresh interval will *always* be blocked to query the DB, so performance is suffering even in the absence of high QPS.

Reply:
N.B. a `Watch`-based implementation (w/ a refresher task) would not be holding the lock during IO, so lock contention would be very cheap -- the lock would only be held long enough to clone the `Arc` inside (for readers), or to replace the `Arc` (for the writer task).

Reply:
Going through the `Watch` docs again, this simpler approach probably should have occurred to me in the first place :laughing:. We could just have a background task that only locks the cache after it's queried the DB.

Refresh task:
```rust
        let refresh_task_handle = spawn(async move {
            loop {
                sleep(REFRESH_INTERVAL_DURATION).await;
                let new_values = datastore.get_values().await;

                {
                    let mut values = values_mutex.lock().await;
                    if new_values != *values {
                        *values = new_values;
                    }
                }
            }
        });
```

The cache itself is still `Mutex<Vec<HpkeConfig>>` (I'll also spend some time coaxing it into a `Mutex<Arc<Vec` to avoid extraneous allocations).

In my isolated experiment I was able to coax this up to 30k QPS before seeing even 1ms of latency due to lock contention, and cache updates don't block readers for the duration of the tx. As a bonus we can downgrade (upgrade?) to a `std::sync::Mutex` which tokio claims provides better performance, since we're not holding locks across awaits.

How does this approach look @branlwyd?

Reply:
LGTM :) (I'm not sure we save much by checking if `new_values != *values`, but I don't object to it -- I suppose we might end up tracking fewer `Arc`s overall.)

Reply:
Indeed. That has the regrettable side affect of turning a few constructors `async` (who says Rust doesn't have function coloring?).

This happens even if we want to use the RwLock method to block readers until the cache is filled. We have to create a write-locked mutex, which AFAICT requires the function to be async. We can't send a `std::sync::RwLock`'s guard across futures, and we have to `.await` to acquire a `tokio::sync::RwLock`.

Reply:
Opened https://github.com/divviup/janus/issues/1639 to address the concurrency error we saw in `task_aggregator_for()`.